// react-lightning-design-system/lib/scripts/Alert.js
'use strict';

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _Button = require('./Button');

var _Button2 = _interopRequireDefault(_Button);

var _Icon = require('./Icon');

var _Icon2 = _interopRequireDefault(_Icon);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Alert = function (_React$Component) {
  (0, _inherits3.default)(Alert, _React$Component);

  function Alert() {
    (0, _classCallCheck3.default)(this, Alert);
    return (0, _possibleConstructorReturn3.default)(this, (0, _getPrototypeOf2.default)(Alert).apply(this, arguments));
  }

  (0, _createClass3.default)(Alert, [{
    key: 'render',
    value: function render() {
      var _props = this.props;
      var children = _props.children;
      var type = _props.type;
      var alt = _props.alt;
      var icon = _props.icon;
      var iconSize = _props.iconSize;
      var props = (0, _objectWithoutProperties3.default)(_props, ['children', 'type', 'alt', 'icon', 'iconSize']);

      var typeClassName = type && alertTypes.indexOf(type) >= 0 ? 'slds-theme--' + type : null;
      var alertClassNames = (0, _classnames2.default)('slds-notify', 'slds-notify--alert', 'slds-theme--alert-texture');
      return _react2.default.createElement(
        'div',
        (0, _extends3.default)({ className: alertClassNames, role: 'alert' }, props),
        alt ? _react2.default.createElement(
          'span',
          { className: 'slds-assistive-text' },
          alt
        ) : undefined,
        _react2.default.createElement(_Button2.default, { type: 'icon-inverse', icon: 'close', iconSize: 'small', alt: 'Close' }),
        _react2.default.createElement(
          'h2',
          null,
          icon ? _react2.default.createElement(_Icon2.default, { icon: icon, iconSize: iconSize }) : undefined,
          children
        )
      );
    }
  }]);
  return Alert;
}(_react2.default.Component);

exports.default = Alert;

var ALERT_TYPES = ['info', 'success', 'warning', 'error'];

Alert.propTypes = {
  type: _react.PropTypes.oneOf(ALERT_TYPES),
  icon: _react.PropTypes.string,
  iconSize: _react.PropTypes.string,
  children: _react.PropTypes.node
};
