// react-lightning-design-system/lib/scripts/FormElement.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _typeof2 = require('babel-runtime/helpers/typeof');

var _typeof3 = _interopRequireDefault(_typeof2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _uuid = require('uuid');

var _uuid2 = _interopRequireDefault(_uuid);

var _util = require('./util');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var FormElement = function (_React$Component) {
  (0, _inherits3.default)(FormElement, _React$Component);

  function FormElement(props) {
    (0, _classCallCheck3.default)(this, FormElement);

    var _this = (0, _possibleConstructorReturn3.default)(this, (0, _getPrototypeOf2.default)(FormElement).call(this, props));

    _this.state = { id: props.id || 'form-element-' + (0, _uuid2.default)() };
    (0, _util.registerStyle)('dropdown', [['.react-slds-dropdown-wrapper', '{ position: relative; }'], ['.slds-modal .react-slds-dropdown-wrapper', '{ position: absolute; }']]);
    return _this;
  }

  (0, _createClass3.default)(FormElement, [{
    key: 'renderFormElement',
    value: function renderFormElement(props) {
      var className = props.className;
      var label = props.label;
      var required = props.required;
      var error = props.error;
      var totalCols = props.totalCols;
      var pprops = (0, _objectWithoutProperties3.default)(props, ['className', 'label', 'required', 'error', 'totalCols']);

      var inputId = props.id || this.state.id;
      if (typeof totalCols === 'number') {
        return _react2.default.createElement(
          'label',
          { className: className },
          label ? _react2.default.createElement(
            'small',
            { className: 'slds-form-element__helper' },
            label
          ) : null,
          this.props.children
        );
      }
      var formElementClassNames = (0, _classnames2.default)('slds-form-element', {
        'slds-has-error': error,
        'slds-is-required': required
      });
      var errorMessage = error ? typeof error === 'string' ? error : (typeof error === 'undefined' ? 'undefined' : (0, _typeof3.default)(error)) === 'object' ? error.message : undefined : undefined;
      return _react2.default.createElement(
        'div',
        { className: formElementClassNames },
        label ? _react2.default.createElement(
          'label',
          { className: 'slds-form-element__label', htmlFor: inputId },
          label
        ) : null,
        _react2.default.createElement(
          'div',
          { className: className },
          this.props.children
        ),
        errorMessage ? _react2.default.createElement(
          'span',
          { className: 'slds-form-element__help' },
          errorMessage
        ) : undefined
      );
    }
  }, {
    key: 'render',
    value: function render() {
      var _props = this.props;
      var className = _props.className;
      var cols = _props.cols;
      var dropdown = _props.dropdown;
      var props = (0, _objectWithoutProperties3.default)(_props, ['className', 'cols', 'dropdown']);

      var colNum = cols || 1;
      var ctrlClassNames = (0, _classnames2.default)('slds-form-element__control', typeof props.totalCols === 'number' ? 'slds-size--' + colNum + '-of-' + props.totalCols : null, className);
      if (dropdown) {
        return _react2.default.createElement(
          'div',
          { className: ctrlClassNames, style: { position: 'static' } },
          this.renderFormElement(props),
          _react2.default.createElement(
            'div',
            { className: 'react-slds-dropdown-wrapper' },
            dropdown
          )
        );
      }

      return this.renderFormElement((0, _extends3.default)({ className: ctrlClassNames }, props));
    }
  }]);
  return FormElement;
}(_react2.default.Component);

exports.default = FormElement;


FormElement.propTypes = {
  id: _react.PropTypes.string,
  dropdown: _react.PropTypes.element,
  className: _react.PropTypes.string,
  required: _react.PropTypes.bool,
  error: _react.PropTypes.bool,
  errorMessage: _react.PropTypes.string,
  cols: _react.PropTypes.number,
  children: _react.PropTypes.element
};

FormElement.isFormElement = true;
