// react-lightning-design-system/lib/scripts/Grid.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Col = exports.Row = undefined;

var _keys = require('babel-runtime/core-js/object/keys');

var _keys2 = _interopRequireDefault(_keys);

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Grid = function (_Component) {
  (0, _inherits3.default)(Grid, _Component);

  function Grid() {
    (0, _classCallCheck3.default)(this, Grid);
    return (0, _possibleConstructorReturn3.default)(this, (0, _getPrototypeOf2.default)(Grid).apply(this, arguments));
  }

  (0, _createClass3.default)(Grid, [{
    key: 'render',
    value: function render() {
      var _props = this.props;
      var className = _props.className;
      var frame = _props.frame;
      var children = _props.children;
      var props = (0, _objectWithoutProperties3.default)(_props, ['className', 'frame', 'children']);

      var gridClassNames = (0, _classnames2.default)(className, 'slds-grid', 'slds-grid--vertical', frame ? 'slds-grid--frame' : null);
      return _react2.default.createElement(
        'div',
        (0, _extends3.default)({ className: gridClassNames }, props),
        children
      );
    }
  }]);
  return Grid;
}(_react.Component);

Grid.propTypes = {
  className: _react.PropTypes.string,
  frame: _react.PropTypes.bool,
  children: _react.PropTypes.node
};

var Row = exports.Row = function (_Component2) {
  (0, _inherits3.default)(Row, _Component2);

  function Row() {
    (0, _classCallCheck3.default)(this, Row);
    return (0, _possibleConstructorReturn3.default)(this, (0, _getPrototypeOf2.default)(Row).apply(this, arguments));
  }

  (0, _createClass3.default)(Row, [{
    key: 'renderColumn',
    value: function renderColumn(colProps, child) {
      /* eslint-disable no-use-before-define */
      if (child.type !== Col) {
        return _react2.default.createElement(
          Col,
          colProps,
          child
        );
      }

      var childProps = (0, _keys2.default)(colProps).reduce(function (cprops, key) {
        cprops[key] = child.props[key] || colProps[key];
        return cprops;
      }, {});
      return _react2.default.cloneElement(child, childProps);
    }
  }, {
    key: 'render',
    value: function render() {
      var _props2 = this.props;
      var className = _props2.className;
      var align = _props2.align;
      var nowrap = _props2.nowrap;
      var nowrapSmall = _props2.nowrapSmall;
      var nowrapMedium = _props2.nowrapMedium;
      var nowrapLarge = _props2.nowrapLarge;
      var cols = _props2.cols;
      var colsSmall = _props2.colsSmall;
      var colsMedium = _props2.colsMedium;
      var colsLarge = _props2.colsLarge;
      var children = _props2.children;
      var props = (0, _objectWithoutProperties3.default)(_props2, ['className', 'align', 'nowrap', 'nowrapSmall', 'nowrapMedium', 'nowrapLarge', 'cols', 'colsSmall', 'colsMedium', 'colsLarge', 'children']);

      var rowClassNames = (0, _classnames2.default)(className, 'slds-grid', align ? 'slds-grid--align-' + align : null, nowrap ? 'slds-nowrap' : 'slds-wrap', nowrapSmall ? 'slds-nowrap--small' : null, nowrapMedium ? 'slds-nowrap--medium' : null, nowrapLarge ? 'slds-nowrap--large' : null);
      var totalCols = cols || function () {
        var cnt = 0;
        _react2.default.Children.forEach(children, function (child) {
          return cnt += child.props.cols || 1;
        });
        return cnt;
      }();
      var colProps = {
        totalCols: totalCols,
        totalColsSmall: colsSmall || totalCols,
        totalColsMedium: colsMedium || totalCols,
        totalColsLarge: colsLarge || totalCols
      };
      return _react2.default.createElement(
        'div',
        (0, _extends3.default)({ className: rowClassNames }, props),
        _react2.default.Children.map(children, this.renderColumn.bind(this, colProps))
      );
    }
  }]);
  return Row;
}(_react.Component);

var ROW_ALIGNS = ['center', 'space', 'spread'];

Row.propTypes = {
  className: _react.PropTypes.string,
  align: _react.PropTypes.oneOf(ROW_ALIGNS),
  nowrap: _react.PropTypes.bool,
  nowrapSmall: _react.PropTypes.bool,
  nowrapMedium: _react.PropTypes.bool,
  nowrapLarge: _react.PropTypes.bool,
  cols: _react.PropTypes.number,
  colsSmall: _react.PropTypes.number,
  colsMedium: _react.PropTypes.number,
  colsLarge: _react.PropTypes.number,
  children: _react.PropTypes.node
};

function adjustCols(colNum, large) {
  if (colNum > 6) {
    return large ? 12 : 6;
  }
  return colNum;
}

var Col = exports.Col = function (_Component3) {
  (0, _inherits3.default)(Col, _Component3);

  function Col() {
    (0, _classCallCheck3.default)(this, Col);
    return (0, _possibleConstructorReturn3.default)(this, (0, _getPrototypeOf2.default)(Col).apply(this, arguments));
  }

  (0, _createClass3.default)(Col, [{
    key: 'render',
    value: function render() {
      var _props3 = this.props;
      var className = _props3.className;
      var padded = _props3.padded;
      var align = _props3.align;
      var noFlex = _props3.noFlex;
      var order = _props3.order;
      var orderSmall = _props3.orderSmall;
      var orderMedium = _props3.orderMedium;
      var orderLarge = _props3.orderLarge;
      var cols = _props3.cols;
      var colsSmall = _props3.colsSmall;
      var colsMedium = _props3.colsMedium;
      var colsLarge = _props3.colsLarge;
      var totalCols = _props3.totalCols;
      var totalColsSmall = _props3.totalColsSmall;
      var totalColsMedium = _props3.totalColsMedium;
      var totalColsLarge = _props3.totalColsLarge;
      var children = _props3.children;
      var props = (0, _objectWithoutProperties3.default)(_props3, ['className', 'padded', 'align', 'noFlex', 'order', 'orderSmall', 'orderMedium', 'orderLarge', 'cols', 'colsSmall', 'colsMedium', 'colsLarge', 'totalCols', 'totalColsSmall', 'totalColsMedium', 'totalColsLarge', 'children']);

      var rowClassNames = (0, _classnames2.default)(className, padded ? 'slds-col--padded' + (/^(medium|large)$/.test(padded) ? '-' + padded : '') : 'slds-col', align ? 'slds-align-' + align : null, noFlex ? 'slds-no-flex' : null, order ? 'slds-order--' + order : null, orderSmall ? 'slds-small-order--' + orderSmall : null, orderMedium ? 'slds-medium-order--' + orderMedium : null, orderLarge ? 'slds-large-order--' + orderLarge : null, cols && totalCols ? 'slds-size--' + cols + '-of-' + adjustCols(totalCols, true) : null, colsSmall && totalColsSmall ? 'slds-small-size--' + colsSmall + '-of-' + adjustCols(totalColsSmall) : null, colsMedium && totalColsMedium ? 'slds-medium-size--' + colsMedium + '-of-' + adjustCols(totalColsMedium) : null, colsLarge && totalColsMedium ? 'slds-large-size--' + colsLarge + '-of-' + adjustCols(totalColsLarge, true) : null);
      return _react2.default.createElement(
        'div',
        (0, _extends3.default)({ className: rowClassNames }, props),
        children
      );
    }
  }]);
  return Col;
}(_react.Component);

var COL_ALIGNS = ['top', 'medium', 'bottom'];

Col.propTypes = {
  className: _react.PropTypes.string,
  padded: _react.PropTypes.oneOfType([_react.PropTypes.bool, _react.PropTypes.string]),
  align: _react.PropTypes.oneOf(COL_ALIGNS),
  noFlex: _react.PropTypes.bool,
  order: _react.PropTypes.number,
  orderSmall: _react.PropTypes.number,
  orderMedium: _react.PropTypes.number,
  orderLarge: _react.PropTypes.number,
  nowrap: _react.PropTypes.bool,
  cols: _react.PropTypes.number,
  colsSmall: _react.PropTypes.number,
  colsMedium: _react.PropTypes.number,
  colsLarge: _react.PropTypes.number,
  totalCols: _react.PropTypes.number,
  totalColsSmall: _react.PropTypes.number,
  totalColsMedium: _react.PropTypes.number,
  totalColsLarge: _react.PropTypes.number,
  children: _react.PropTypes.node
};

exports.default = Grid;
