// react-lightning-design-system/lib/scripts/Notification.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Toast = exports.Alert = undefined;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _Button = require('./Button');

var _Button2 = _interopRequireDefault(_Button);

var _Icon = require('./Icon');

var _Icon2 = _interopRequireDefault(_Icon);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var NOTIFICATION_TYPES = ['alert', 'toast'];

var NOTIFICATION_LEVELS = ['info', 'success', 'warning', 'error'];

var Notification = function (_Component) {
  (0, _inherits3.default)(Notification, _Component);

  function Notification() {
    (0, _classCallCheck3.default)(this, Notification);
    return (0, _possibleConstructorReturn3.default)(this, (0, _getPrototypeOf2.default)(Notification).apply(this, arguments));
  }

  (0, _createClass3.default)(Notification, [{
    key: 'render',
    value: function render() {
      var _props = this.props;
      var className = _props.className;
      var type = _props.type;
      var level = _props.level;
      var alt = _props.alt;
      var _props$alertTexture = _props.alertTexture;
      var alertTexture = _props$alertTexture === undefined ? true : _props$alertTexture;
      var icon = _props.icon;
      var _props$iconSize = _props.iconSize;
      var iconSize = _props$iconSize === undefined ? 'small' : _props$iconSize;
      var onClose = _props.onClose;
      var children = _props.children;
      var props = (0, _objectWithoutProperties3.default)(_props, ['className', 'type', 'level', 'alt', 'alertTexture', 'icon', 'iconSize', 'onClose', 'children']);

      var typeClassName = type && NOTIFICATION_TYPES.indexOf(type) >= 0 ? 'slds-notify--' + type : null;
      var levelClassName = type && NOTIFICATION_LEVELS.indexOf(level) >= 0 ? 'slds-theme--' + level : null;
      var alertClassNames = (0, _classnames2.default)(className, 'slds-notify', typeClassName, levelClassName, alertTexture ? 'slds-theme--alert-texture' : null);
      return _react2.default.createElement(
        'div',
        (0, _extends3.default)({ className: alertClassNames, role: 'alert' }, props),
        alt ? _react2.default.createElement(
          'span',
          { className: 'slds-assistive-text' },
          alt
        ) : undefined,
        onClose ? _react2.default.createElement(_Button2.default, { className: 'slds-notify__close', type: 'icon-inverse',
          icon: 'close', iconSize: 'small', alt: 'Close',
          onClick: onClose
        }) : undefined,
        icon ? _react2.default.createElement(_Icon2.default, { className: 'slds-m-right--x-small',
          icon: icon, size: iconSize,
          fillColor: 'none', textColor: level === 'warning' ? 'default' : null
        }) : undefined,
        children
      );
    }
  }]);
  return Notification;
}(_react.Component);

exports.default = Notification;


Notification.propTypes = {
  type: _react.PropTypes.oneOf(NOTIFICATION_TYPES).isRequired,
  className: _react.PropTypes.string,
  level: _react.PropTypes.oneOf(NOTIFICATION_LEVELS),
  alt: _react.PropTypes.string,
  icon: _react.PropTypes.string,
  iconSize: _react.PropTypes.string,
  children: _react.PropTypes.node,
  onClose: _react.PropTypes.func
};

var propTypes = (0, _extends3.default)({}, Notification.propTypes);
delete propTypes.type;

var Alert = exports.Alert = function (_Component2) {
  (0, _inherits3.default)(Alert, _Component2);

  function Alert() {
    (0, _classCallCheck3.default)(this, Alert);
    return (0, _possibleConstructorReturn3.default)(this, (0, _getPrototypeOf2.default)(Alert).apply(this, arguments));
  }

  (0, _createClass3.default)(Alert, [{
    key: 'render',
    value: function render() {
      return _react2.default.createElement(Notification, (0, _extends3.default)({}, this.props, { type: 'alert' }));
    }
  }]);
  return Alert;
}(_react.Component);

Alert.propTypes = propTypes;

var Toast = exports.Toast = function (_Component3) {
  (0, _inherits3.default)(Toast, _Component3);

  function Toast() {
    (0, _classCallCheck3.default)(this, Toast);
    return (0, _possibleConstructorReturn3.default)(this, (0, _getPrototypeOf2.default)(Toast).apply(this, arguments));
  }

  (0, _createClass3.default)(Toast, [{
    key: 'render',
    value: function render() {
      return _react2.default.createElement(Notification, (0, _extends3.default)({}, this.props, { type: 'toast' }));
    }
  }]);
  return Toast;
}(_react.Component);

Toast.propTypes = propTypes;
