// react-lightning-design-system/lib/scripts/Picklist.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PicklistItem = undefined;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _FormElement = require('./FormElement');

var _FormElement2 = _interopRequireDefault(_FormElement);

var _Icon = require('./Icon');

var _Icon2 = _interopRequireDefault(_Icon);

var _DropdownMenu = require('./DropdownMenu');

var _DropdownMenu2 = _interopRequireDefault(_DropdownMenu);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Picklist = function (_React$Component) {
  (0, _inherits3.default)(Picklist, _React$Component);

  function Picklist(props) {
    (0, _classCallCheck3.default)(this, Picklist);

    var _this = (0, _possibleConstructorReturn3.default)(this, (0, _getPrototypeOf2.default)(Picklist).call(this, props));

    _this.state = { opened: props.defaultOpened, value: props.defaultValue };
    return _this;
  }

  (0, _createClass3.default)(Picklist, [{
    key: 'componentDidUpdate',
    value: function componentDidUpdate(prevProps, prevState) {
      if (this.props.onValueChange && prevState.value !== this.state.value) {
        this.props.onValueChange(this.state.value, prevState.value);
      }
    }
  }, {
    key: 'onClick',
    value: function onClick() {
      var _this2 = this;

      this.setState({ opened: !this.state.opened });
      setTimeout(function () {
        _this2.focusToTargetItemEl();
      }, 10);
    }
  }, {
    key: 'onPicklistItemClick',
    value: function onPicklistItemClick(item, e) {
      var _this3 = this;

      this.setState({ value: item.value });
      if (this.props.onChange) {
        this.props.onChange(e, item.value);
      }
      if (this.props.onSelect) {
        this.props.onSelect(item);
      }
      setTimeout(function () {
        _this3.setState({ opened: false });
        if (_this3.props.onComplete) {
          _this3.props.onComplete();
        }
        var picklistButtonEl = _reactDom2.default.findDOMNode(_this3.refs.picklistButton);
        if (picklistButtonEl) {
          picklistButtonEl.focus();
        }
      }, 200);
      e.preventDefault();
      e.stopPropagation();
    }
  }, {
    key: 'onPicklistClose',
    value: function onPicklistClose() {
      var picklistButtonEl = _reactDom2.default.findDOMNode(this.refs.picklistButton);
      picklistButtonEl.focus();
      this.setState({ opened: false });
    }
  }, {
    key: 'onBlur',
    value: function onBlur() {
      var _this4 = this;

      setTimeout(function () {
        if (!_this4.isFocusedInComponent()) {
          _this4.setState({ opened: false });
          if (_this4.props.onBlur) {
            _this4.props.onBlur();
          }
          if (_this4.props.onComplete) {
            _this4.props.onComplete();
          }
        }
      }, 10);
    }
  }, {
    key: 'onKeydown',
    value: function onKeydown(e) {
      var _this5 = this;

      if (e.keyCode === 40) {
        // down
        e.preventDefault();
        e.stopPropagation();
        if (!this.state.opened) {
          this.setState({ opened: true });
          setTimeout(function () {
            _this5.focusToTargetItemEl();
          }, 10);
        } else {
          this.focusToTargetItemEl();
        }
      } else if (e.keyCode === 27) {
        // ESC
        e.preventDefault();
        e.stopPropagation();
        this.setState({ opened: false });
        if (this.props.onComplete) {
          this.props.onComplete();
        }
      }
      if (this.props.onKeyDown) {
        this.props.onKeyDown(e);
      }
    }
  }, {
    key: 'getSelectedValue',
    value: function getSelectedValue() {
      var _props = this.props;
      var defaultValue = _props.defaultValue;
      var value = _props.value;

      return typeof value !== 'undefined' ? value : typeof this.state.value !== 'undefined' ? this.state.value : defaultValue;
    }
  }, {
    key: 'getSelectedItemLabel',
    value: function getSelectedItemLabel() {
      var selectedValue = this.getSelectedValue();
      var selected = null;
      _react2.default.Children.forEach(this.props.children, function (item) {
        if (item.props.value === selectedValue) {
          selected = item.props.label || item.props.children;
        }
      });
      return selected || this.props.selectedText;
    }
  }, {
    key: 'isFocusedInComponent',
    value: function isFocusedInComponent() {
      var rootEl = _reactDom2.default.findDOMNode(this);
      var targetEl = document.activeElement;
      while (targetEl && targetEl !== rootEl) {
        targetEl = targetEl.parentNode;
      }
      return !!targetEl;
    }
  }, {
    key: 'focusToTargetItemEl',
    value: function focusToTargetItemEl() {
      var dropdownEl = _reactDom2.default.findDOMNode(this.refs.dropdown);
      var firstItemEl = dropdownEl.querySelector('.slds-is-selected > .react-slds-menuitem[tabIndex]') || dropdownEl.querySelector('.react-slds-menuitem[tabIndex]');
      if (firstItemEl) {
        firstItemEl.focus();
      }
    }
  }, {
    key: 'renderPicklist',
    value: function renderPicklist(props) {
      var className = props.className;
      var id = props.id;
      var menuSize = props.menuSize;
      var children = props.children;
      var pprops = (0, _objectWithoutProperties3.default)(props, ['className', 'id', 'menuSize', 'children']);

      var picklistClassNames = (0, _classnames2.default)(className, 'slds-picklist');
      return _react2.default.createElement(
        'div',
        { className: picklistClassNames, 'aria-expanded': this.state.opened },
        _react2.default.createElement(
          'button',
          (0, _extends3.default)({ id: id, ref: 'picklistButton', className: 'slds-picklist__label slds-button slds-button--neutral',
            type: 'button', 'aria-haspopup': true }, pprops, {
            onClick: this.onClick.bind(this),
            onBlur: this.onBlur.bind(this),
            onKeyDown: this.onKeydown.bind(this)
          }),
          _react2.default.createElement(
            'span',
            { className: 'slds-truncate' },
            this.getSelectedItemLabel() || _react2.default.createElement(
              'span',
              null,
              ' '
            )
          ),
          _react2.default.createElement(_Icon2.default, { icon: 'down' })
        )
      );
    }
  }, {
    key: 'renderDropdown',
    value: function renderDropdown() {
      var _props2 = this.props;
      var menuSize = _props2.menuSize;
      var children = _props2.children;

      return this.state.opened ? _react2.default.createElement(
        _DropdownMenu2.default,
        { ref: 'dropdown', size: menuSize,
          onMenuItemClick: this.onPicklistItemClick.bind(this),
          onMenuClose: this.onPicklistClose.bind(this)
        },
        _react2.default.Children.map(children, this.renderPicklistItem.bind(this))
      ) : _react2.default.createElement('div', { ref: 'dropdown' });
    }
  }, {
    key: 'renderPicklistItem',
    value: function renderPicklistItem(item) {
      var selected = item.props.value === this.getSelectedValue();
      var onBlur = this.onBlur.bind(this);
      return _react2.default.cloneElement(item, { selected: selected, onBlur: onBlur });
    }
  }, {
    key: 'render',
    value: function render() {
      var _props3 = this.props;
      var label = _props3.label;
      var required = _props3.required;
      var error = _props3.error;
      var totalCols = _props3.totalCols;
      var cols = _props3.cols;
      var props = (0, _objectWithoutProperties3.default)(_props3, ['label', 'required', 'error', 'totalCols', 'cols']);

      var dropdown = this.renderDropdown();
      var formElemProps = { id: props.id, label: label, required: required, error: error, totalCols: totalCols, cols: cols, dropdown: dropdown };
      return _react2.default.createElement(
        _FormElement2.default,
        formElemProps,
        this.renderPicklist(props)
      );
    }
  }]);
  return Picklist;
}(_react2.default.Component);

exports.default = Picklist;


Picklist.propTypes = {
  className: _react.PropTypes.string,
  label: _react.PropTypes.string,
  required: _react.PropTypes.bool,
  error: _react.PropTypes.oneOfType([_react.PropTypes.bool, _react.PropTypes.string, _react.PropTypes.shape({
    message: _react.PropTypes.string
  })]),
  name: _react.PropTypes.string,
  value: _react.PropTypes.any,
  defaultValue: _react.PropTypes.any,
  selectedText: _react.PropTypes.string,
  defaultOpened: _react.PropTypes.bool,
  onChange: _react.PropTypes.func,
  onValueChange: _react.PropTypes.func,
  onSelect: _react.PropTypes.func,
  onComplete: _react.PropTypes.func,
  onKeyDown: _react.PropTypes.func,
  onBlur: _react.PropTypes.func,
  totalCols: _react.PropTypes.number,
  cols: _react.PropTypes.number,
  // TODO: Such props should be taken from component
  menuSize: _react.PropTypes.string,
  children: _react.PropTypes.node
};

Picklist.isFormElement = true;

var PicklistItem = exports.PicklistItem = function (_React$Component2) {
  (0, _inherits3.default)(PicklistItem, _React$Component2);

  function PicklistItem() {
    (0, _classCallCheck3.default)(this, PicklistItem);
    return (0, _possibleConstructorReturn3.default)(this, (0, _getPrototypeOf2.default)(PicklistItem).apply(this, arguments));
  }

  (0, _createClass3.default)(PicklistItem, [{
    key: 'render',
    value: function render() {
      var _props4 = this.props;
      var label = _props4.label;
      var selected = _props4.selected;
      var value = _props4.value;
      var children = _props4.children;
      var props = (0, _objectWithoutProperties3.default)(_props4, ['label', 'selected', 'value', 'children']);

      return _react2.default.createElement(
        _DropdownMenu.DropdownMenuItem,
        (0, _extends3.default)({ icon: selected ? 'check' : 'none', role: 'menuitemradio', selected: selected }, props),
        label || children
      );
    }
  }]);
  return PicklistItem;
}(_react2.default.Component);

PicklistItem.propTypes = {
  label: _react.PropTypes.oneOfType([_react.PropTypes.string, _react.PropTypes.number]),
  selected: _react.PropTypes.bool,
  value: _react.PropTypes.any,
  children: _react.PropTypes.node
};
