// react-lightning-design-system/lib/scripts/Spinner.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _util = require('./util');

var _util2 = _interopRequireDefault(_util);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Spinner = function (_React$Component) {
  (0, _inherits3.default)(Spinner, _React$Component);

  function Spinner() {
    (0, _classCallCheck3.default)(this, Spinner);
    return (0, _possibleConstructorReturn3.default)(this, (0, _getPrototypeOf2.default)(Spinner).apply(this, arguments));
  }

  (0, _createClass3.default)(Spinner, [{
    key: 'render',
    value: function render() {
      var _props = this.props;
      var className = _props.className;
      var size = _props.size;
      var type = _props.type;
      var alt = _props.alt;
      var props = (0, _objectWithoutProperties3.default)(_props, ['className', 'size', 'type', 'alt']);

      var spinnerClassNames = (0, _classnames2.default)(className, 'slds-spinner--' + size);
      var spinnerImgName = type === 'brand' ? 'slds_spinner_brand' : type === 'inverse' ? 'slds_spinner_inverse' : 'slds_spinner';
      return _react2.default.createElement(
        'div',
        (0, _extends3.default)({ className: spinnerClassNames }, props),
        _react2.default.createElement('img', { src: _util2.default.getAssetRoot() + '/images/spinners/' + spinnerImgName + '.gif', alt: alt })
      );
    }
  }]);
  return Spinner;
}(_react2.default.Component);

exports.default = Spinner;


Spinner.propTypes = {
  className: _react.PropTypes.string,
  type: _react.PropTypes.string,
  size: _react.PropTypes.string,
  alt: _react.PropTypes.string
};
