// react-lightning-design-system/lib/scripts/Tabs.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _util = require('./util');

var _DropdownButton = require('./DropdownButton');

var _DropdownButton2 = _interopRequireDefault(_DropdownButton);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Tabs = function (_React$Component) {
  (0, _inherits3.default)(Tabs, _React$Component);

  function Tabs(props) {
    (0, _classCallCheck3.default)(this, Tabs);

    var _this = (0, _possibleConstructorReturn3.default)(this, (0, _getPrototypeOf2.default)(Tabs).call(this, props));

    _this.state = {};
    (0, _util.registerStyle)('tab-menu', [['.slds-tabs__item.react-slds-tab-with-menu', '{ position: relative !important; overflow: visible !important; }'], ['.slds-tabs__item.react-slds-tab-with-menu > .react-slds-tab-item-inner', '{ overflow: hidden }'], ['.slds-tabs__item.react-slds-tab-with-menu > .react-slds-tab-item-inner > a', '{ padding-right: 2rem; }'], ['.react-slds-tab-menu', '{ position: absolute; top: 0; right: 0; visibility: hidden }'], ['.react-slds-tab-menu button', '{ height: 3rem; line-height: 3rem; width: 2rem; }'], ['.slds-tabs__item.slds-active .react-slds-tab-menu', '.slds-tabs__item:hover .react-slds-tab-menu', '{ visibility: visible }']]);
    return _this;
  }

  (0, _createClass3.default)(Tabs, [{
    key: 'componentDidUpdate',
    value: function componentDidUpdate() {
      if (this.state.focusTab) {
        var el = _reactDom2.default.findDOMNode(this.refs.activeTab);
        if (el) {
          el.focus();
        }
        /* eslint-disable react/no-did-update-set-state */
        this.setState({ focusTab: false });
      }
    }
  }, {
    key: 'onTabClick',
    value: function onTabClick(tabKey) {
      if (this.props.onSelect) {
        this.props.onSelect(tabKey);
      }
      // Uncontrolled
      this.setState({ activeKey: tabKey, focusTab: true });
    }
  }, {
    key: 'onTabKeyDown',
    value: function onTabKeyDown(tabKey, e) {
      var _this2 = this;

      if (e.keyCode === 37 || e.keyCode === 39) {
        (function () {
          // left/right cursor key
          var idx = 0;
          var tabKeys = [];
          _react2.default.Children.forEach(_this2.props.children, function (tab, i) {
            tabKeys.push(tab.props.eventKey);
            if (tabKey === tab.props.eventKey) {
              idx = i;
            }
          });
          var dir = e.keyCode === 37 ? -1 : 1;
          var activeIdx = (idx + dir + tabKeys.length) % tabKeys.length;
          var activeKey = tabKeys[activeIdx];
          _this2.onTabClick(activeKey);
          e.preventDefault();
          e.stopPropagation();
        })();
      }
    }
  }, {
    key: 'renderTabNav',
    value: function renderTabNav(type, tabs) {
      var _this3 = this;

      var activeKey = typeof this.props.activeKey !== 'undefined' ? this.props.activeKey : typeof this.state.activeKey !== 'undefined' ? this.state.activeKey : this.props.defaultActiveKey;
      var tabNavClassName = 'slds-tabs--' + type + '__nav';
      return _react2.default.createElement(
        'ul',
        { className: tabNavClassName, role: 'tablist' },
        _react2.default.Children.map(tabs, function (tab) {
          var _tab$props = tab.props;
          var title = _tab$props.title;
          var eventKey = _tab$props.eventKey;
          var menu = _tab$props.menu;
          var menuIcon = _tab$props.menuIcon;
          var menuItems = tab.props.menuItems;

          menuItems = menu ? menu.props.children : menuItems;
          var menuProps = menu ? menu.props : {};
          var isActive = eventKey === activeKey;
          var tabItemClassName = (0, _classnames2.default)('slds-tabs__item', 'slds-tabs--' + type + '__item', 'slds-text-heading---label', { 'slds-active': isActive }, { 'react-slds-tab-with-menu': menu || menuItems });
          var tabLinkClassName = 'slds-tabs--' + type + '__link';
          return _react2.default.createElement(
            'li',
            { className: tabItemClassName, role: 'presentation' },
            _react2.default.createElement(
              'span',
              { className: 'react-slds-tab-item-inner' },
              _react2.default.createElement(
                'a',
                { className: tabLinkClassName,
                  onClick: _this3.onTabClick.bind(_this3, eventKey),
                  onKeyDown: _this3.onTabKeyDown.bind(_this3, eventKey),
                  role: 'tab',
                  ref: isActive ? 'activeTab' : null,
                  tabIndex: isActive ? 0 : -1,
                  'aria-selected': isActive
                },
                title
              ),
              menuItems ? _this3.renderTabMenu(menuIcon, menuItems, menuProps) : null
            )
          );
        })
      );
    }
  }, {
    key: 'renderTabMenu',
    value: function renderTabMenu() {
      var menuIcon = arguments.length <= 0 || arguments[0] === undefined ? 'down' : arguments[0];
      var menuItems = arguments.length <= 1 || arguments[1] === undefined ? [] : arguments[1];
      var menuProps = arguments.length <= 2 || arguments[2] === undefined ? {} : arguments[2];

      return _react2.default.createElement(
        _DropdownButton2.default,
        (0, _extends3.default)({ className: 'react-slds-tab-menu', icon: menuIcon, type: 'icon-bare', iconSize: 'small', nubbinTop: true }, menuProps),
        menuItems
      );
    }
  }, {
    key: 'renderTabPanel',
    value: function renderTabPanel(tab) {
      var activeKey = this.props.activeKey || this.state.activeKey || this.props.defaultActiveKey;
      var eventKey = tab.props.eventKey;

      var isActive = eventKey === activeKey;
      return _react2.default.cloneElement(tab, { active: isActive });
    }
  }, {
    key: 'render',
    value: function render() {
      var _props = this.props;
      var className = _props.className;
      var children = _props.children;
      var props = (0, _objectWithoutProperties3.default)(_props, ['className', 'children']);

      var type = this.props.type === 'scoped' ? 'scoped' : 'default';
      var tabsClassNames = (0, _classnames2.default)(className, 'slds-tabs--' + type);
      return _react2.default.createElement(
        'div',
        { className: tabsClassNames },
        this.renderTabNav(type, children),
        _react2.default.Children.map(children, this.renderTabPanel.bind(this))
      );
    }
  }]);
  return Tabs;
}(_react2.default.Component);

exports.default = Tabs;


var TAB_TYPES = ['default', 'scoped'];

Tabs.propTypes = {
  className: _react.PropTypes.string,
  type: _react.PropTypes.oneOf(TAB_TYPES),
  defaultActiveKey: _react.PropTypes.any,
  activeKey: _react.PropTypes.any,
  onSelect: _react.PropTypes.func,
  children: _react.PropTypes.node
};
